README.md
# File Embedder

Contributor: mehdi jalili
Tags: pdf, embed, File Embedder, viewer, block
Requires at least: 5.8
Tested up to: 6.1
Requires PHP: 7.4
Stable tag: 1.0.0
License: GPL v2+
License URI: https://www.gnu.org/licenses/gpl-2.0.html

File Embedder helps you to Embed PDF from the Media Library or via easily.

## Description
File Embedder helps you to Embed PDF from the Media Library or via easily.

Inspired by [Embed PDF](https://wordpress.org/plugins/embed-pdf-viewer/)


Development on [GitHub](https://github.com/mehdi-jalili/file-embedder). Pull requests are welcome against the `develop` branch.

Pull requests are welcome against the `develop` branch.

## Screenshots
![File Embedder block](./.wordpress-org/screenshot-1.png "File Embedder block")

File Embedder block

![Select PDF](./.wordpress-org/screenshot-2.png "Select PDF")

Select PDF


![PDF block and attributes](./.wordpress-org/screenshot-3.png "PDF block and attributes")

PDF block and attributes


CHANGES.md
#### [unreleased]

#### 1.0
* initial commit


wp-embedder.php
<?php
/**
 * wp Embedder
 *
 * @author  Mehdi Jalili
 * @license GPL-2.0+
 * @link    https://github.com/mehdi-jalili/wp-embedder
 * @package wp-embedder
 */

/**
 * Plugin Name:       Wp Embedder
 * Plugin URI:        https://github.com/mehdi-jalili/wp-embedder
 * Description:       Embed PDF Files from the Media Library, Google Doc Viewer and or everywhere.
 * Author:            Mehdi Jalili
 * Author URI:        https://github.com/mehdi-jalili
 * Version:           1.0.0
 * License:           GPLv2+
 * Domain Path:       /languages
 * Text Domain:       wp-embedder
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.html
 * GitHub Plugin URI: https://github.com/afragen/wp-embedder
 * Requires PHP:      7.4
 * Requires at least: 5.8
 */

/**
 * Exit if called directly.
 */
if ( ! defined( 'WPINC' ) ) {
	die;
}

add_filter( 'media_send_to_editor', [ Wp_Embedder::instance(), 'embed_pdf_media_editor' ], 20, 2 );
wp_embed_register_handler(
	'owp-embedder',
	'#(^(https?)\:\/\/.+\.pdf$)#i',
	[
		Wp_Embedder::instance(),
		'owp-embedder',
	]
);

add_action(
	'init',
	function () {
		load_plugin_textdomain( 'wp-embedder' );
		wp_set_script_translations( 'wp-embedder-scripts', 'wp-embedder' );

		// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.NoExplicitVersion
		wp_enqueue_style(
			'wp-embedder',
			plugins_url( 'css/wp-embedder.css', __FILE__ ),
			[],
			false,
			'screen'
		);
	}
);
add_action( 'init', [ Wp_Embedder::instance(), 'register_block' ] );

/**
 * Class Wp_Embedder
 */
class Wp_Embedder {
	/**
	 * For singleton.
	 *
	 * @var bool
	 */
	private static $instance = false;

	/**
	 * Create singleton.
	 *
	 * @return bool
	 */
	public static function instance() {
		if ( false === self::$instance ) {
			self::$instance = new self();
		}

		return self::$instance;
	}

	/**
	 * Register block.
	 *
	 * @return void
	 */
	public function register_block() {
		if ( ! function_exists( 'register_block_type' ) ) {
			return;
		}

		// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.NoExplicitVersion
		wp_register_script(
			'wp-embedder',
			plugins_url( 'blocks/build/index.js', __FILE__ ),
			[ 'wp-i18n', 'wp-blocks', 'wp-block-editor', 'wp-element', 'wp-components', 'wp-compose', 'wp-blob' ],
			false,
			true
		);

		register_block_type(
			'wp-embedder/pdf',
			[
				'editor_script' => 'wp-embedder',
			]
		);
	}

	/**
	 * Insert URL to PDF from Media Library, then render as oEmbed.
	 *
	 * @param string  $html an href link to the media.
	 * @param integer $id   post_id.
	 *
	 * @return string
	 */
	public function embed_pdf_media_editor( $html, $id ) {
		$post = get_post( $id );
		if ( 'application/pdf' !== $post->post_mime_type ) {
			return $html;
		}

		return $post->guid . "\n\n";
	}

	/**
	 * Create oEmbed code.
	 *
	 * @param array  $matches Regex matches.
	 * @param array  $atts    array of media height/width.
	 * @param string $url     URI for media file.
	 *
	 * @return string
	 */
	public function owp_embedder( $matches, $atts, $url ) {
		$attachment_id = $this->get_attachment_id_by_url( $url );
		if ( ! empty( $attachment_id ) ) {
			$post = get_post( $this->get_attachment_id_by_url( $url ) );
		} else {
			/*
			 * URL is from outside of the Media Library.
			 */
			$post                 = new WP_Post( new stdClass() );
			$post->guid           = $matches[0];
			$post->post_mime_type = 'application/pdf';
			$post->post_name      = preg_replace( '/\.pdf$/', '', basename( $matches[0] ) );
		}

		return $this->create_output( $post, $atts );
	}

	/**
	 * Create output for Google Doc Viewer and href link to file.
	 *
	 * @param \WP_Post     $post Current post.
	 * @param array|string $atts array of media height/width or
	 *                           href to media library asset.
	 *
	 * @return bool|string
	 */
	private function create_output( WP_Post $post, $atts = [] ) {
		if ( 'application/pdf' !== $post->post_mime_type ) {
			return $atts;
		}

		$default = [
			'height'      => 500,
			'width'       => 800,
			'title'       => $post->post_title,
			'description' => $post->post_content,
		];

		/*
		 * Ensure $atts isn't the href.
		 */
		$atts = is_array( $atts ) ? $atts : [];

		if ( isset( $atts['height'] ) ) {
			$atts['height'] = ( $atts['height'] / 2 );
		}
		$atts = array_merge( $default, $atts );

		/**
		 * Filter PDF attributes.
		 *
		 * @since 1.6.0
		 * @param  array $atts Array of PDF attributes.
		 * @return array $atts
		 */
		$atts = apply_filters( 'embed_pdf_viewer_pdf_attributes', $atts );

		// Fix title or create from filename.
		$atts['title']       = empty( $atts['title'] )
			? ucwords( preg_replace( '/(-|_)/', ' ', $post->post_name ) )
			: ucwords( preg_replace( '/(-|_)/', ' ', $atts['title'] ) );
		$atts['description'] = empty( $atts['description'] ) ? $atts['title'] : $atts['description'];

		$iframe_fallback  = '<iframe class="wp-embedder" src="https://docs.google.com/viewer?url=' . rawurlencode( $post->guid );
		$iframe_fallback .= '&amp;embedded=true" frameborder="0" ';
		$iframe_fallback .= 'style="height:' . $atts['height'] . 'px;width:' . $atts['width'] . 'px;" ';
		$iframe_fallback .= 'title="' . $atts['description'] . '"></iframe>' . "\n";

		$object  = '<object class="wp-embedder" data="' . $post->guid;
		$object .= '#scrollbar=1&toolbar=1"';
		$object .= 'type="application/pdf" ';
		$object .= 'height=' . $atts['height'] . ' width=' . $atts['width'] . ' ';
		$object .= 'title="' . $atts['description'] . '"> ';
		$object .= '</object>';

		$embed  = '<figure>';
		$embed .= $object . $iframe_fallback;
		$embed .= '<p><a href="' . $post->guid . '" title="' . $atts['description'] . '">' . $atts['title'] . '</a></p>';
		$embed .= '</figure>';

		return $embed;
	}

	/**
	 * Get attachment id by url. Thanks Pippin.
	 *
	 * @link  https://pippinsplugins.com/retrieve-attachment-id-from-image-url/
	 *
	 * @param string $url URI of attachment.
	 *
	 * @return mixed
	 */
	private function get_attachment_id_by_url( $url ) {
		global $wpdb;
		// phpcs:ignore WordPress.DB
		$attachment = $wpdb->get_col( $wpdb->prepare( "SELECT ID FROM $wpdb->posts WHERE guid='%s';", $url ) );

		if ( empty( $attachment ) ) {
			return null;
		}

		return $attachment[0];
	}
}


blocks/src/index.js
import classnames from 'classnames';
import icons from './icons.js';

const { __ } = wp.i18n;
const { registerBlockType, getBlockDefaultClassName } = wp.blocks;
const { RichText, MediaPlaceholder, MediaUpload, InspectorControls, BlockControls, BlockAlignmentToolbar } = wp.blockEditor;
const { Fragment } = wp.element;
const { withNotices, Button, TextControl, TextareaControl, PanelBody, ToolbarGroup, ToolbarButton, ResizableBox } = wp.components;
const { withState } = wp.compose;
const { isBlobURL } = wp.blob;

const isExternalPDF = (id, url) => url && !id && !isBlobURL(url);

const renderEmbed = (props) => {
	const { attributes: { title, description, url, width, height, align } } = props;
	const style = { width, height };
	const myClassName = getBlockDefaultClassName('file-embedder/pdf');

	if (undefined === url || !url) {
		return null;
	}

	return (
		<figure className={`${myClassName}__content-wrapper align${align}`}>
			<object
				className="file-embedder"
				data={url + '#scrollbar=1&toolbar=1'}
				type="application/pdf"
				height={style.height}
				width={style.width}
				title={description}
			>
			</object>
			<iframe
				className="file-embedder"
				src={'https://docs.google.com/viewer?url=' + encodeURIComponent(url) + '&embedded=true'}
				frameBorder="0"
				height={style.height}
				width={style.width}
				title={description}
			>
			</iframe>
		</figure>
	);
};

const renderEdit = (props) => {
	const { attributes: { id, title, description, url, width, height, align }, setAttributes, isEditing, hasError, setState, className, media, noticeUI, noticeOperations, toggleSelection, isRTL } = props;
	const isExternal = isExternalPDF(id, url);

	function updateAttribute(key) {
		return function (value) {
			const attr = {};
			attr[key] = value;
			setAttributes(attr);
		}
	}

	function onSelectFile(media) {
		if (media && media.url) {
			setState({
				hasError: false,
			});
			setAttributes({
				url: media.url,
				id: media.id,
				title: media.title,
				description: media.description,
			});
		}
	}

	function toggleIsEditing() {
		setState({
			isEditing: !isEditing,
		});
	}

	if (undefined === url || !url || hasError || isEditing) {
		return (
			<MediaPlaceholder
				icon={icons.pdf}
				labels={{
					title: __('PDF', 'file-embedder'),
					instructions: __('Drag a PDF, upload a new one or select a PDF from your library.', 'file-embedder'),
				}}
				onSelect={onSelectFile}
				onSelectURL={updateAttribute('url')}
				notices={props.noticeUI}
				onError={props.noticeOperations.createErrorNotice}
				accept='application/pdf'
				allowedTypes={['application/pdf']}
			/>
		);
	}


	const classes = classnames(className, {
		'is-transient': isBlobURL(url),
	});


	let showRightHandle = false;
	let showLeftHandle = false;


	if (align === 'center') {

		showRightHandle = true;
		showLeftHandle = true;
	} else if (isRTL) {

		if (align === 'left') {
			showRightHandle = true;
		} else {
			showLeftHandle = true;
		}
	} else {

		if (align === 'right') {
			showLeftHandle = true;
		} else {
			showRightHandle = true;
		}
	}

	return (
		<Fragment>
			<InspectorControls>
				<PanelBody title={__('File Embedder', 'file-embedder')} initialOpen={true}>
					<div>
						<TextareaControl
							label={__('Long Description (optional)', 'file-embedder')}
							value={undefined === description ? '' : description}
							onChange={updateAttribute('description')}
							help={__('Long Description used for `title` tag and accessibility.', 'file-embedder')}
						/>
						<TextControl
							type="number"
							min={20}
							label={__('Width', 'file-embedder')}
							value={undefined === width ? FileEmbedder.attributes.width.default : width}
							onChange={updateAttribute('width')}
						/>
						<TextControl
							type="number"
							label={__('Height', 'file-embedder')}
							value={undefined === height ? fileEmbedder.attributes.height.default : height}
							min={1}
							onChange={updateAttribute('height')}
						/>
					</div>
				</PanelBody>
			</InspectorControls>

			<BlockControls>
				<BlockAlignmentToolbar
					value={align}
					onChange={updateAttribute('align')}
					controls={['left', 'center', 'right']}
				/>
				<ToolbarGroup>
					{isExternal && (
						<ToolbarButton
							className="components-icon-button"
							label={__('Edit PDF', 'file-embedder')}
							onClick={toggleIsEditing}
							icon="edit"
						/>
					)}
					{!isExternal && (
						<MediaUpload
							onSelect={onSelectFile}
							value={id}
							render={({ open }) => (
								<Button
									className="components-toolbar__control"
									label={__('Edit PDF', 'file-embedder')}
									onClick={open}
									icon="edit"
								/>
							)}
						/>
					)}
				</ToolbarGroup>
			</BlockControls>

			<div className={classes}>
				<ResizableBox
					size={
						width && height ? {
							width,
							height,
						} : undefined
					}
					lockAspectRatio
					enable={{
						top: false,
						right: showRightHandle,
						bottom: true,
						left: showLeftHandle,
					}}
					onResizeStart={() => {
						toggleSelection(false);
					}}
					onResizeStop={(event, direction, elt, delta) => {
						setAttributes({
							width: parseInt(width + delta.width, 10),
							height: parseInt(height + delta.height, 10),
						});
						toggleSelection(true);
					}}
				>
					{renderEmbed(props)}
				</ResizableBox>
			</div>
		</Fragment>
	);
};

let embedPDFViewer = registerBlockType('file-embedder/pdf', {
	title: __('PDF', 'file-embedder'),
	icon: icons.pdf,
	category: 'embed',
	attributes: {
		id: { type: 'number', },
		title: { type: 'string' },
		description: {
			type: 'string',
			default: '',
		},
		url: { type: 'string', },
		width: {
			type: 'string',
			default: 600,
		},
		height: {
			type: 'string',
			default: 600,
		},
		align: {
			type: 'string',
			default: 'center',
		},
		supports: {
			align: ['left', 'center', 'right'],
		},
	},

	getEditWrapperProps(attributes) {
		const { align, width } = attributes;
		if ('left' === align || 'center' === align || 'right' === align) {
			return { 'data-align': align, 'data-resized': !!width };
		}
	},

	edit: withNotices(withState({ isEditing: false, hasError: false })(renderEdit)),
	save: renderEmbed,
});


file-embedder.php
<?php
/**
 * File Embedder
 *
 * @author  Mehdi Jalili
 * @license GPL-2.0+
 * @link    https://github.com/afragen/file-embedder
 * @package file-embedder
 */

/**
 * Plugin Name:       File Embedder
 * Plugin URI:        https://github.com/afragen/file-embedder
 * Description:       File Embedder helps you to Embed PDF from the Media Library or via easily.
 * Author:            Mehdi Jalili
 * Author URI:        https://github.com/mehdi-jalili
 * Version:           1.0.0
 * License:           GPLv2+
 * Domain Path:       /languages
 * Text Domain:       file-embedder
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.html
 * GitHub Plugin URI: https://github.com/mehdi-jalili/file-embedder
 * Requires PHP:      7.4
 * Requires at least: 5.8
 */

/**
 * Exit if called directly.
 */
if ( ! defined( 'WPINC' ) ) {
	die;
}

add_filter( 'media_send_to_editor', [ Flle_Embedder::instance(), 'embed_pdf_media_editor' ], 20, 2 );
wp_embed_register_handler(
	'flle_embedder',
	'#(^(https?)\:\/\/.+\.pdf$)#i',
	[
		Flle_Embedder::instance(),
		'flle_embedder',
	]
);
add_action(
	'init',
	function () {
		load_plugin_textdomain( 'flle-embedder' );
		wp_set_script_translations( 'flle-embedder-scripts', 'flle-embedder' );

		// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.NoExplicitVersion
		wp_enqueue_style(
			'flle-embedder',
			plugins_url( 'css/flle-embedder.css', __FILE__ ),
			[],
			false,
			'screen'
		);
	}
);
add_action( 'init', [ Flle_Embedder::instance(), 'register_block' ] );

/**
 * Class Flle_Embedder
 */
class Flle_Embedder {
	/**
	 * For singleton.
	 *
	 * @var bool
	 */
	private static $instance = false;

	/**
	 * Create singleton.
	 *
	 * @return bool
	 */
	public static function instance() {
		if ( false === self::$instance ) {
			self::$instance = new self();
		}

		return self::$instance;
	}

	/**
	 * Register block.
	 *
	 * @return void
	 */
	public function register_block() {
		if ( ! function_exists( 'register_block_type' ) ) {
			return;
		}

		// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.NoExplicitVersion
		wp_register_script(
			'flle-embedder',
			plugins_url( 'blocks/build/index.js', __FILE__ ),
			[ 'wp-i18n', 'wp-blocks', 'wp-block-editor', 'wp-element', 'wp-components', 'wp-compose', 'wp-blob' ],
			false,
			true
		);

		register_block_type(
			'flle-embedder/pdf',
			[
				'editor_script' => 'flle-embedder',
			]
		);
	}

	/**
	 * Insert URL to PDF from Media Library, then render as oEmbed.
	 *
	 * @param string  $html an href link to the media.
	 * @param integer $id   post_id.
	 *
	 * @return string
	 */
	public function embed_pdf_media_editor( $html, $id ) {
		$post = get_post( $id );
		if ( 'application/pdf' !== $post->post_mime_type ) {
			return $html;
		}

		return $post->guid . "\n\n";
	}

	/**
	 * Create oEmbed code.
	 *
	 * @param array  $matches Regex matches.
	 * @param array  $atts    array of media height/width.
	 * @param string $url     URI for media file.
	 *
	 * @return string
	 */
	public function flle_embedder( $matches, $atts, $url ) {
		$attachment_id = $this->get_attachment_id_by_url( $url );
		if ( ! empty( $attachment_id ) ) {
			$post = get_post( $this->get_attachment_id_by_url( $url ) );
		} else {
			/*
			 * URL is from outside of the Media Library.
			 */
			$post                 = new WP_Post( new stdClass() );
			$post->guid           = $matches[0];
			$post->post_mime_type = 'application/pdf';
			$post->post_name      = preg_replace( '/\.pdf$/', '', basename( $matches[0] ) );
		}

		return $this->create_output( $post, $atts );
	}

	/**
	 * Create output for Google Doc Viewer and href link to file.
	 *
	 * @param \WP_Post     $post Current post.
	 * @param array|string $atts array of media height/width or
	 *                           href to media library asset.
	 *
	 * @return bool|string
	 */
	private function create_output( WP_Post $post, $atts = [] ) {
		if ( 'application/pdf' !== $post->post_mime_type ) {
			return $atts;
		}

		$default = [
			'height'      => 500,
			'width'       => 800,
			'title'       => $post->post_title,
			'description' => $post->post_content,
		];

		/*
		 * Ensure $atts isn't the href.
		 */
		$atts = is_array( $atts ) ? $atts : [];

		if ( isset( $atts['height'] ) ) {
			$atts['height'] = ( $atts['height'] / 2 );
		}
		$atts = array_merge( $default, $atts );

		/**
		 * Filter PDF attributes.
		 *
		 * @since 1.6.0
		 * @param  array $atts Array of PDF attributes.
		 * @return array $atts
		 */
		$atts = apply_filters( 'embed_pdf_viewer_pdf_attributes', $atts );

		// Fix title or create from filename.
		$atts['title']       = empty( $atts['title'] )
			? ucwords( preg_replace( '/(-|_)/', ' ', $post->post_name ) )
			: ucwords( preg_replace( '/(-|_)/', ' ', $atts['title'] ) );
		$atts['description'] = empty( $atts['description'] ) ? $atts['title'] : $atts['description'];

		$iframe_fallback  = '<iframe class="flle-embedder" src="https://docs.google.com/viewer?url=' . rawurlencode( $post->guid );
		$iframe_fallback .= '&amp;embedded=true" frameborder="0" ';
		$iframe_fallback .= 'style="height:' . $atts['height'] . 'px;width:' . $atts['width'] . 'px;" ';
		$iframe_fallback .= 'title="' . $atts['description'] . '"></iframe>' . "\n";

		$object  = '<object class="flle-embedder" data="' . $post->guid;
		$object .= '#scrollbar=1&toolbar=1"';
		$object .= 'type="application/pdf" ';
		$object .= 'height=' . $atts['height'] . ' width=' . $atts['width'] . ' ';
		$object .= 'title="' . $atts['description'] . '"> ';
		$object .= '</object>';

		$embed  = '<figure>';
		$embed .= $object . $iframe_fallback;
		$embed .= '<p><a href="' . $post->guid . '" title="' . $atts['description'] . '">' . $atts['title'] . '</a></p>';
		$embed .= '</figure>';

		return $embed;
	}

	/**
	 * Get attachment id by url. Thanks Pippin.
	 *
	 * @link  https://pippinsplugins.com/retrieve-attachment-id-from-image-url/
	 *
	 * @param string $url URI of attachment.
	 *
	 * @return mixed
	 */
	private function get_attachment_id_by_url( $url ) {
		global $wpdb;
		// phpcs:ignore WordPress.DB
		$attachment = $wpdb->get_col( $wpdb->prepare( "SELECT ID FROM $wpdb->posts WHERE guid='%s';", $url ) );

		if ( empty( $attachment ) ) {
			return null;
		}

		return $attachment[0];
	}
}
